packages/protocol/contracts/stability/PairManager.sol

```diff
@@ -0,0 +1,290 @@
+  // Tracks the total fraction of reserve stable that has been allocated to existing virtual pairs.
+  mapping(address => FixidityLib.Fraction) public allocatedStableFractions;
+
+  /* ==================== Constructor ==================== */
+
+  /**
+   * @notice Sets initialized == true on implementation contracts.
+   * @param test Set to true to skip implementation initialization.
+   */
+  constructor(bool test) public Initializable(test) {}
+
+  /**
+   * @notice Allows the contract to be upgradable via the proxy.
+   * @param _broker The address of the broker contract.
+   * @param registryAddress The address of the Celo registry.
+   */
+  function initilize(address _broker, address registryAddress) external initializer {
+    _transferOwnership(msg.sender);
+    setRegistry(registryAddress);
+    setBroker(_broker);
+  }
+
+  /* ==================== Modifiers ==================== */
+
+  modifier onlyBroker() {
+    require(msg.sender == broker, "Caller is not the Broker");
+    _;
+  }
+
+  /* ==================== View Functions ==================== */
+
+  function getPair(bytes32 pairId) public view returns (Pair memory pair) {
+    require(isPair[pairId], "A pair with the specified id does not exist");
+    pair = pairs[pairId];
+  }
+
+  /* ==================== Mutative Functions ==================== */
+
+  /**
+   * @notice Sets the address of the broker contract.
+   * @param _broker The new address of the broker contract.
+   */
+  function setBroker(address _broker) public onlyOwner {
+    require(_broker != address(0), "Broker address must be set");
+    broker = _broker;
+    emit BrokerUpdated(_broker);
+  }
+
+  /**
+   * @notice Creates a new pair using the given parameters.
+   * @param pair The information required to create the pair.
+   * @return pairId The id of the newly created pair.
+   */
+  function createPair(Pair calldata pair) external onlyOwner returns (bytes32 pairId) {
+    require(address(pair.mentoExchange) != address(0), "Mento exchange must be set");
+    require(address(pair.stableAsset) != address(0), "Stable asset must be set");
+    require(address(pair.collateralAsset) != address(0), "Collateral asset must be set");
+
+    Pair memory pairInfo = pair;
+
+    pairId = keccak256(
+      abi.encodePacked(
+        IERC20Metadata(pairInfo.stableAsset).symbol(),
+        IERC20Metadata(pairInfo.collateralAsset).symbol(),
+        pairInfo.mentoExchange.name()
+      )
+    );
+    require(!isPair[pairId], "A pair with the specified assets and exchange exists");
+
+    validatePairInfo(pairInfo);
+
+    // TODO: getUpdatedBuckets Function sig should change. May not need amounts
+    // but instead pass minSupplyForStableBucketCap & stableBucketMaxFraction
+    (uint256 tokenInBucket, uint256 tokenOutBucket) = pairInfo.mentoExchange.getUpdatedBuckets(
+      pairInfo.stableAsset,
+      pairInfo.collateralAsset,
+      0,
+      0,
+      pairId
+    );
```

Comment:
We can discuss this when we look at the MentoExchange implementations but in my head exchange functions would be `pure view` and staticcalled so all information needed to make a decision would be passed in, they shouldn't query the pairManager, but we can burn that bridge when we get to it ( 😂 )

Reply:
Agreed. Before I handed off to nina the signature took in the whole pair as it basically needs everything. But yeah I agree, 🔥 🌉